spring-test/src/main/java/org/springframework/test/context/jdbc/Sql.java

```diff
@@ -148,12 +153,23 @@
 	 */
 	SqlConfig config() default @SqlConfig;
 
-
 	/**
 	 * Enumeration of <em>phases</em> that dictate when SQL scripts are executed.
 	 */
 	enum ExecutionPhase {
 
+		/**
+		 * The configured SQL scripts and statements will be executed
+		 * once <em>before</em> any test method is run.
+		 */
+		BEFORE_TEST_CLASS,
```

Comment:
It's a bit strange to have a method level annotation that can dictate something that applies to the whole class.

Reply:
I'm aware that it's not optimal. As mentioned in the PR description, there are more consistency issues. My question is: Is this acceptable? Or would it be better to introduce a new annotation like `@SqlPerClass` (might not be the best name)? The upside would be that the problems around `MergeMode` would be gone, too.

Reply:
[DirtiesContext](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/test/annotation/DirtiesContext.html) has a similar design, but exposes two elements:`classMode` and `methodMode`. Still, nothing prevents me from annotating a method with `@DirtiesContext(classMode = BEFORE_CLASS)` which has no effect.

Reply:
Before @aahlenst posted those last two comments, I was also going to point out that there is a precedent for this, namely `@DirtiesContext`.

I agree that it is a bit strange for it to be possible to specify a class-level setting at the method level that simply gets ignored; however, I am not in favor of introducing yet another SQL-related annotation for this purpose. We already have quite a few. And I don't think it necessarily warrants an additional enum and annotation attribute.

As it stands in the proposal, the addition of the two new enum constants in `ExecutionPhase` makes it impossible for a user to configure conflicting phases for a given `@Sql` declaration, and that's a good thing.

FWIW, I have never heard any complaints from users about the `ClassMode` and `MethodMode` dichotomy in `@DirtiesContext`, and the documentation clearly states that setting an inappropriate mode "has no meaning" (i.e., will be ignored).

In light of that, I am OK with the PR as-is.

Reply:
Can we at lease throw an exception if you're using the class-level mode on a method?

Reply:
Yes, we can definitely throw an exception for inappropriate user configuration.

`IllegalArgumentException` sounds reasonable to me.

Reply:
We throw now.

Just thinking aloud: If `@Sql` has this, wouldn't it be nice if `@DirtiesContext` had it, too? Or would that be too disruptive years after the introduction?

Reply:
> We throw now.

Thanks

> Just thinking aloud: If `@Sql` has this, wouldn't it be nice if `@DirtiesContext` had it, too? Or would that be too disruptive years after the introduction?

`@Sql` is different because it has a single enum-based `executionPhase` attribute, which means we can easily detect a user configuration error.

Whereas, `@DirtiesContext` has two mutually exclusive enum-based attributes: `classMode` and `methodMode`, and each of those has a `default` value. Thus, we cannot actually detect a user configuration error.

The only way to be able to detect a user configuration error for `@DirtiesContext` would be to introduce new `DEFAULT` enum constants in `ClassMode` and `MethodMode` and use those as the `default` attribute values going forward.

However, that would be a breaking change for any users who had inadvertently explicitly supplied the previous default values for those attributes.

Plus, it would make things a bit more cumbersome.

In light of the above, I don't think we should make any changes to `@DirtiesContext` in this regard this late in the game.